sql,logictest: validate all descriptors after logic test completion

spaskob opened the pull request:
After each logic test completes we will validate all descriptors
present in the cluster and fail if problems are discovered. After
that we drop the test db with `CASCADE` and repeat the same
validation.

Fixes #53754.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53810)
<!-- Reviewable:end -->

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -4175,6 +4176,37 @@ may increase either contention or retry errors, or both.`,
+				if err != nil {
+					return nil, err
+				}
+				dg := catalogkv.NewOneLevelUncachedDescGetter(ctx.Txn, ctx.Codec)
```
I feel like we should be using a cached desc getter at this point

ajwerner reviewed (commented):
@spaskob and @rohany see https://github.com/cockroachdb/cockroach/pull/53904 for how I think I'd do this. We'll also want to invoke the logic to validate the various namespace entries that @spaskob has been adding. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-:-MGJcfBs7plxqI40RxKi:b1srwxy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @lucy-zhang, and @spaskob)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
It seemed easier to just type up the implementation than to try to explain it due to the various complications on extending existing abstractions. 

I'd like us to extract these virtual tables and the descriptor handling abstractions around them into catalog on the 21.1 cycle which should allow us to strengthen and unify some of this code with code elsewhere. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-:-MGJcqwmEHZnwECM6U9f:b-wmjdv2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @lucy-zhang, and @spaskob)



<!-- Sent from Reviewable.io -->

rohany commented:
I'm a fan  #53904  in doing this. It gives us alot more flexibility of things to do without having to expose strange interfaces in the eval context etc

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/sem/builtins/builtins.go, line 4190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MGCKK78o8ifTW1o1Dpg-r1-4190:-MIwY2sbEbkVPyvAHcO9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f36045e404aa9942519d7ec1085c4c4a85e5e7/pkg/sql/sem/builtins/builtins.go#L4190)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I feel like we should be using a cached desc getter at this point
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob commented:
This is now using `crdb_internal.invalid_objects`.
PTAL

Commit:
logictest: validate all descriptors at test completion

After each logic test completes we will validate all descriptors
present in the cluster and fail if problems are discovered. After
that we drop the test db with `CASCADE` and repeat the same
validation.

Fixes #53754.

Release note: none.

thoszhang reviewed (commented):
> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, @rohany, and @spaskob)

___
*[pkg/sql/logictest/logic.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDntpf5zVVYSPq4wo9:-MJDntpf5zVVYSPq4woA:boverxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2696)):*
> ```Go
> 	}
> 
> 	t.setUser("root")
> ```

We should pull all this into a separate function, I think.
___
*[pkg/sql/logictest/logic.go, line 2736 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDnlsJEaog9TgNe9Bb:-MJDnlsJEaog9TgNe9Bc:bwdgi8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2736)):*
> ```Go
> 	}
> 
> 	if _, err := t.db.Exec("SET sql_safe_updates=false;"); err != nil {
> ```

Is this just so we can drop the current database if it's `test`?
___
*[pkg/sql/logictest/logic.go, line 2739 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDnHJYCPOdph-sgU0O:-MJDnHJYCPOdph-sgU0P:boubx8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2739)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if _, err := t.db.Exec("DROP DATABASE test CASCADE"); err != nil {
> ```

Can we just do `DROP DATABASE IF EXISTS`?

Also, I think at some point we'll want to drop all created databases in this test, not just the one we started with. Can we add a TODO here?
___
*[pkg/sql/logictest/logic.go, line 2741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDms8o1KdgqPZ36-Of:-MJDms8o1KdgqPZ36-Og:b-2qju6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2741)):*
> ```Go
> 	if _, err := t.db.Exec("DROP DATABASE test CASCADE"); err != nil {
> 		if !strings.Contains(err.Error(), `database "test" does not exist`) {
> 			t.Fatal(err)
> ```

It'd be good to wrap this error with an explanation so it's more obvious what a failure indicates.


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/logictest/logic.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDntpf5zVVYSPq4wo9:-MJMPQcj0WfE5YMcQage:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2696)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We should pull all this into a separate function, I think.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/logic.go, line 2736 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDnlsJEaog9TgNe9Bb:-MJMHVOQ2L4aFzZGuUlF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2736)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is this just so we can drop the current database if it's `test`?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/logic.go, line 2739 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDnHJYCPOdph-sgU0O:-MJMHZulD4sv_blFQ8Pr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2739)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Can we just do `DROP DATABASE IF EXISTS`?

Also, I think at some point we'll want to drop all created databases in this test, not just the one we started with. Can we add a TODO here?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/logic.go, line 2741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJDms8o1KdgqPZ36-Of:-MJMMlsXChvZJTWTZ2xT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6208481deae92d01db1288b456c5b27a4e70dd3/pkg/sql/logictest/logic.go#L2741)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

It'd be good to wrap this error with an explanation so it's more obvious what a failure indicates.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, @rohany, and @spaskob)

___
*[pkg/sql/logictest/logic.go, line 2696 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSa4qM9fuJPcOuT8u5:-MJSa4qN09QKZSY8q1TV:bm2k131) ([raw file](https://github.com/cockroachdb/cockroach/blob/159920d9aa7b72112e6ce0f25b1cc610cdc115c3/pkg/sql/logictest/logic.go#L2696)):*
> ```Go
> 
> 	validate := func() (string, error) {
> 		rows, err := t.db.Query(`SELECT * FROM "".crdb_internal.invalid_objects ORDER BY id`)
> ```

I wonder if there are cases where a logic test has switched connections (through the `user`) subcommand and then leaves a transaction open. One way we could deal with that would be to run this at high priority:

```
require.NoError(t, crdb.ExecuteTx(ctx, t.db, *gosql.TxnOptions{}, func(txn *sql.Tx) error {
    if _, err := txn.Exec("SET TRANSACTION PRIORITY HIGH"); err != nil {
        return err
    }
    ...
}))
```
___
*[pkg/sql/logictest/logic.go, line 2729 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSYOmU9hQt-0qxc_vI:-MJSYOmU9hQt-0qxc_vJ:b-355ihu) ([raw file](https://github.com/cockroachdb/cockroach/blob/159920d9aa7b72112e6ce0f25b1cc610cdc115c3/pkg/sql/logictest/logic.go#L2729)):*
> ```Go
> 
> 	// TODO(lucy): we should really drop all created databases in this test, not
> 	// `just the one we started with.
> ```

nit: \`
___
*[pkg/sql/sem/builtins/builtins.go, line 4186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSZeP30e0LTRCfETzj:-MJSZeP41Bg5Ui0outWw:b-ve7crf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f36045e404aa9942519d7ec1085c4c4a85e5e7/pkg/sql/sem/builtins/builtins.go#L4186)):*
> ```Go
> 			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 				stmt := "SELECT id FROM system.descriptor ORDER BY id"
> 				rows, err := ctx.InternalExecutor.Query(ctx.Context, "fetch-descriptors", nil, stmt)
> ```

My sense is that this should use the user's transaction, otherwise it's exposed to deadlock in cases where the transaction has read other stuff and just generally will not be properly tracked for deadlock detection. 


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/logictest/logic.go, line 2696 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSa4qM9fuJPcOuT8u5:-MJVd0JW1eZMFiGwAljk:b-rb2e4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/159920d9aa7b72112e6ce0f25b1cc610cdc115c3/pkg/sql/logictest/logic.go#L2696)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I wonder if there are cases where a logic test has switched connections (through the `user`) subcommand and then leaves a transaction open. One way we could deal with that would be to run this at high priority:

```
require.NoError(t, crdb.ExecuteTx(ctx, t.db, *gosql.TxnOptions{}, func(txn *sql.Tx) error {
    if _, err := txn.Exec("SET TRANSACTION PRIORITY HIGH"); err != nil {
        return err
    }
    ...
}))
```
</blockquote></details>

Can you explain which code do you want to wrap with this?
___
*[pkg/sql/sem/builtins/builtins.go, line 4186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSZeP30e0LTRCfETzj:-MJVcl7k2pYuxRwk7U31:b-ugdtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f36045e404aa9942519d7ec1085c4c4a85e5e7/pkg/sql/sem/builtins/builtins.go#L4186)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

My sense is that this should use the user's transaction, otherwise it's exposed to deadlock in cases where the transaction has read other stuff and just generally will not be properly tracked for deadlock detection. 
</blockquote></details>

this file is no longer part of this PR


<!-- Sent from Reviewable.io -->

ajwerner commented:
@spaskob I added code to close the connections associated with other users before validation. The problem was that in some workloads we have some transactions left open on other connections.

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53810)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @lucy-zhang, @rohany, and @spaskob)

___
*[pkg/sql/logictest/logic.go, line 2696 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53810#-MJSa4qM9fuJPcOuT8u5:-MJYYmMvAllF7RMXDeX_:bxgqug6) ([raw file](https://github.com/cockroachdb/cockroach/blob/159920d9aa7b72112e6ce0f25b1cc610cdc115c3/pkg/sql/logictest/logic.go#L2696)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Can you explain which code do you want to wrap with this?
</blockquote></details>

Closing the concurrent connections dealt with this too so skipping. 


<!-- Sent from Reviewable.io -->

spaskob commented:
@ajwerner thanks for your help!
This looks good to go then.

ajwerner commented:
Any time!

bors r+
